packages/feedback/src/screenshot/components/ScreenshotEditor.tsx

```diff
@@ -2,171 +2,372 @@ import type { FeedbackInternalOptions, FeedbackModalIntegration } from '@sentry/
+
+  /**
+   * A reference to the whole dialog (the parent of this component) so that we
+   * can show/hide it and take a clean screenshot of the webpage.
+   */
   dialog: ReturnType<FeedbackModalIntegration['createDialog']>;
+
+  /**
+   * The whole options object.
+   *
+   * Needed to set nonce and id values for editor specific styles
+   */
   options: FeedbackInternalOptions;
 }
 
 interface Props {
   onError: (error: Error) => void;
 }
 
-interface Box {
-  startX: number;
-  startY: number;
-  endX: number;
-  endY: number;
-}
+type MaybeCanvas = HTMLCanvasElement | null;
+type Screenshot = { canvas: HTMLCanvasElement; dpi: number };
 
-interface Rect {
+
+type DrawType = 'highlight' | 'hide' | '';
+interface DrawCommand {
+  type: DrawType;
   x: number;
   y: number;
-  height: number;
-  width: number;
+  h: number;
+  w: number;
+}
+
+function drawRect(command: DrawCommand, ctx: CanvasRenderingContext2D, color: string): void {
+  switch (command.type) {
+    case 'highlight': {
+      // creates a shadow around
+      ctx.shadowColor = 'rgba(0, 0, 0, 0.7)';
+      ctx.shadowBlur = 50;
+
+      // draws a rectangle first with a shadow
+      ctx.fillStyle = color;
+      ctx.fillRect(command.x - 1, command.y - 1, command.w + 2, command.h + 2);
+
+      // cut out the inside of the rectangle
+      ctx.clearRect(command.x, command.y, command.w, command.h);
+
+      break;
+    }
+    case 'hide':
+      ctx.fillStyle = 'rgb(0, 0, 0)';
+      ctx.fillRect(command.x, command.y, command.w, command.h);
+
+      break;
+    default:
+      break;
+  }
 }
 
-const getContainedSize = (img: HTMLCanvasElement): Rect => {
-  const imgClientHeight = img.clientHeight;
-  const imgClientWidth = img.clientWidth;
-  const ratio = img.width / img.height;
-  let width = imgClientHeight * ratio;
-  let height = imgClientHeight;
-  if (width > imgClientWidth) {
-    width = imgClientWidth;
-    height = imgClientWidth / ratio;
+function with2dContext(
+  canvas: MaybeCanvas,
+  options: CanvasRenderingContext2DSettings,
+  callback: (canvas: HTMLCanvasElement, ctx: CanvasRenderingContext2D) => void,
+): void {
+  if (!canvas) {
+    return;
+  }
+  const ctx = canvas.getContext('2d', options);
+  if (!ctx) {
+    return;
   }
-  const x = (imgClientWidth - width) / 2;
-  const y = (imgClientHeight - height) / 2;
-  return { x: x, y: y, width: width, height: height };
-};
+  callback(canvas, ctx);
+}
+
+function paintImage(maybeDest: MaybeCanvas, source: HTMLCanvasElement): void {
+  with2dContext(maybeDest, { alpha: true }, (destCanvas, destCtx) => {
+    destCtx.drawImage(source, 0, 0, source.width, source.height, 0, 0, destCanvas.width, destCanvas.height);
+  });
+}
+
+// Paint the array of drawCommands into a canvas.
+// Assuming this is the canvas foreground, and the background is cleaned.
+function paintForeground(
+  maybeCanvas: MaybeCanvas,
+  strokeColor: string,
+  drawCommands: DrawCommand[],
+  currentDrawCommand: undefined | DrawCommand
```

Comment:
Why does this need both a list of draw commands and the "current" draw command?

Reply:
ah, i guess i can merge the arrays together with some `[...drawCommands, currentDrawCommand]` at the callsite.

it's not going to be crazy perf impacting, but that's a reduction of a parameter and some lines of code, in favor of a new array during mousemove.